data_pipeline/europepmc/europepmc_pipeline.py

```diff
@@ -0,0 +1,194 @@
+import logging
+from datetime import date, timedelta
+from json.decoder import JSONDecodeError
+from typing import Iterable, NamedTuple, Optional, Sequence
+
+import requests
+
+from data_pipeline.europepmc.europepmc_config import (
+    EuropePmcConfig,
+    EuropePmcSourceConfig,
+    EuropePmcStateConfig
+)
+from data_pipeline.utils.collections import iter_batches_iterable
+from data_pipeline.utils.data_store.bq_data_service import (
+    load_given_json_list_data_from_tempdir_to_bq
+)
+from data_pipeline.utils.data_store.s3_data_service import (
+    download_s3_object_as_string_or_file_not_found_error,
+    upload_s3_object
+)
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+class EuropePmcSearchContext(NamedTuple):
+    start_date_str: str
+    end_date_str: str
+
+    def is_empty_period(self):
+        return date.fromisoformat(self.end_date_str) < date.fromisoformat(self.start_date_str)
+
+
+def iter_article_data_from_response_json(
+    response_json: dict
+) -> Iterable[dict]:
+    return response_json['resultList']['result']
+
+
+def get_request_query_for_source_config_and_start_date_str(
+    source_config: EuropePmcSourceConfig,
+    search_context: EuropePmcSearchContext
+) -> str:
+    date_period_str = (
+        f"['{search_context.start_date_str}' TO '{search_context.end_date_str}']"
+    )
+    query = (
+        f"(FIRST_IDATE:{date_period_str}) "
+        + (source_config.search.query or '')
+    ).strip()
+    return query
+
+
+def get_request_params_for_source_config(
+    source_config: EuropePmcSourceConfig,
+    search_context: EuropePmcSearchContext
+) -> dict:
+    return {
+        **(source_config.search.extra_params or {}),
+        'query': get_request_query_for_source_config_and_start_date_str(
+            source_config,
+            search_context
+        ),
+        'format': 'json',
+        'resultType': 'core'
+    }
+
+
+def get_valid_json_from_response(response: requests.Response) -> dict:
+    try:
+        response.raise_for_status()
+        return response.json()
+    except JSONDecodeError:
+        LOGGER.warning('failed to decode json: %r', response.text)
+        raise
+
+
+def get_article_response_json_from_api(
+    source_config: EuropePmcSourceConfig,
+    search_context: EuropePmcSearchContext
+) -> dict:
+    url = source_config.api_url
+    params = get_request_params_for_source_config(
+        source_config,
+        search_context
+    )
+    response = requests.get(url, params=params)
+    return get_valid_json_from_response(response)
+
+
+def get_filtered_article_data(data: dict, fields_to_return: Optional[Sequence[str]]) -> dict:
```

Comment:
To be more clear; maybe the name of the function can be `get_requested_fields_of_the_data`. What do you think?